File: source14.js
// Source14.js - SGR 0501+4516 Magnetar Module (JavaScript Implementation)
// Based on source14.cpp - Master Universal Gravity Equation (MUGE) for SGR 0501+4516
// Time-Reversal Magnetar with enhanced framework including f_TRZ factor
// Includes ALL terms: base gravity, cosmic expansion, magnetic decay, UQFF Ug with f_TRZ,
// Lambda, scaled EM with UA correction, GW, quantum uncertainty, fluid dynamics,
// oscillatory waves, and DM/density perturbations

class MagnetarSGR0501_4516 {
    constructor() {
        this.initializeDefaults();
    }

    initializeDefaults() {
        // Core UQFF parameters (matching C++ implementation)
        this.G = 6.6743e-11;
        this.M = 1.4 * 1.989e30; // 1.4 M_sun in kg
        this.r = 2e4; // 20 km radius (larger than SGR 1745-2900)
        this.H0 = 2.184e-18; // Different Hubble constant
        this.B0 = 1e10; // Initial B field (weaker than SGR 1745-2900)
        this.tau_B = 4000 * 3.15576e7; // B decay timescale (4000 years)
        this.B_crit = 1e11; // Critical B field
        this.Lambda = 1.1e-52; // Cosmological constant
        this.c_light = 3e8; // Speed of light
        this.q_charge = 1.602e-19; // Proton charge
        this.v_surf = 1e6; // Surface velocity
        this.f_TRZ = 0.1; // Time-reversal factor (unique to SGR 0501+4516)
        this.rho_vac_UA = 7.09e-36; // UA vacuum density
        this.rho_vac_SCm = 7.09e-37; // SCm vacuum density
        this.P_init = 5.0; // Initial rotation period (slower than SGR 1745-2900)
        this.tau_Omega = 10000 * 3.15576e7; // Omega decay timescale
        this.scale_EM = 1e-12; // EM scaling factor
        this.proton_mass = 1.673e-27; // Proton mass

        // Full term parameters
        this.hbar = 1.0546e-34; // Reduced Planck's constant
        this.t_Hubble = 13.8e9 * 3.15576e7; // Hubble time in seconds
        this.delta_x = 1e-10; // Position uncertainty
        this.delta_p = this.hbar / this.delta_x; // Momentum uncertainty
        this.integral_psi = 1.0; // Wavefunction integral
        this.rho_fluid = 1e17; // Fluid density
        this.A_osc = 1e10; // Oscillatory amplitude
        this.k_osc = 1.0 / this.r; // Wave number
        this.omega_osc = 2 * Math.PI / this.P_init; // Angular frequency
        this.x_pos = this.r; // Position for oscillation
        this.t_Hubble_gyr = 13.8; // Hubble time in Gyr
        this.M_DM_factor = 0.1; // Dark matter mass fraction
        this.delta_rho_over_rho = 1e-5; // Density perturbation

        this.updateCache();
    }

    updateCache() {
        this.ug1_base = (this.G * this.M) / (this.r * this.r);
    }

    setVariable(varName, newValue) {
        if (this.hasOwnProperty(varName)) {
            this[varName] = newValue;
            if (varName === 'B0') this.B = newValue;
            this.updateCache();
            return true;
        }
        return false;
    }

    getVariable(varName) {
        return this.hasOwnProperty(varName) ? this[varName] : 0.0;
    }

    addToVariable(varName, delta) {
        return this.setVariable(varName, this.getVariable(varName) + delta);
    }

    subtractFromVariable(varName, delta) {
        return this.addToVariable(varName, -delta);
    }

    B_t(t) {
        return this.B0 * Math.exp(-t / this.tau_B);
    }

    Omega_t(t) {
        return (2 * Math.PI / this.P_init) * Math.exp(-t / this.tau_Omega);
    }

    dOmega_dt(t) {
        const omega0 = 2 * Math.PI / this.P_init;
        return omega0 * (-1.0 / this.tau_Omega) * Math.exp(-t / this.tau_Omega);
    }

    compute_Ug(Bt) {
        const Ug1 = this.ug1_base;
        const Ug2 = 0.0;
        const Ug3 = 0.0;
        const Ug4 = Ug1 * (1 - Bt / this.B_crit);
        return (Ug1 + Ug2 + Ug3 + Ug4) * (1 + this.f_TRZ);
    }

    compute_V() {
        return (4.0 / 3.0) * Math.PI * this.r * this.r * this.r;
    }

    compute_g_Magnetar(t) {
        if (t < 0) {
            console.error("Error: Time t must be non-negative.");
            return 0.0;
        }

        const Bt = this.B_t(t);
        const dOdt = this.dOmega_dt(t);

        // Term 1: Base + H0 + B corrections
        const corr_H = 1 + this.H0 * t;
        const corr_B = 1 - Bt / this.B_crit;
        const term1 = this.ug1_base * corr_H * corr_B;

        // Term 2: UQFF Ug with f_TRZ (time-reversal enhancement)
        const term2 = this.compute_Ug(Bt);

        // Term 3: Lambda
        const term3 = (this.Lambda * this.c_light * this.c_light) / 3.0;

        // Term 4: Scaled EM with UA correction
        const cross_vB = this.v_surf * Bt;
        const em_base = (this.q_charge * cross_vB) / this.proton_mass;
        const corr_UA = 1 + (this.rho_vac_UA / this.rho_vac_SCm);
        const term4 = (em_base * corr_UA) * this.scale_EM;

        // Term 5: GW
        const gw_prefactor = (this.G * this.M * this.M) / (Math.pow(this.c_light, 4) * this.r);
        const term5 = gw_prefactor * (dOdt * dOdt);

        // Quantum uncertainty term
        const sqrt_unc = Math.sqrt(this.delta_x * this.delta_p);
        const term_q = (this.hbar / sqrt_unc) * this.integral_psi * (2 * Math.PI / this.t_Hubble);

        // Fluid term (effective acceleration)
        const V = this.compute_V();
        const term_fluid = (this.rho_fluid * V * this.ug1_base) / this.M;

        // Oscillatory terms (real parts)
        const term_osc1 = 2 * this.A_osc * Math.cos(this.k_osc * this.x_pos) * Math.cos(this.omega_osc * t);
        const arg = this.k_osc * this.x_pos - this.omega_osc * t;
        const term_osc2 = (2 * Math.PI / this.t_Hubble_gyr) * this.A_osc * Math.cos(arg);
        const term_osc = term_osc1 + term_osc2;

        // DM and density perturbation term (converted to acceleration)
        const M_dm = this.M * this.M_DM_factor;
        const pert1 = this.delta_rho_over_rho;
        const pert2 = 3 * this.G * this.M / (this.r * this.r * this.r);
        const term_dm_force_like = (this.M + M_dm) * (pert1 + pert2);
        const term_DM = term_dm_force_like / this.M;

        // Total g_Magnetar (all terms summed)
        return term1 + term2 + term3 + term4 + term5 + term_q + term_fluid + term_osc + term_DM;
    }

    printParameters() {
        console.log("SGR 0501+4516 Parameters:");
        console.log(`G: ${this.G}, M: ${this.M}, r: ${this.r}`);
        console.log(`H0: ${this.H0}, B0: ${this.B0}, tau_B: ${this.tau_B}`);
        console.log(`f_TRZ: ${this.f_TRZ}, rho_fluid: ${this.rho_fluid}, M_DM_factor: ${this.M_DM_factor}`);
        console.log(`A_osc: ${this.A_osc}, delta_rho_over_rho: ${this.delta_rho_over_rho}`);
        console.log(`ug1_base: ${this.ug1_base}`);
    }

    exampleAt5000Years() {
        const t_example = 5000 * 3.15576e7;
        return this.compute_g_Magnetar(t_example);
    }

    getParameters() {
        return {
            name: "SGR 0501+4516",
            G: this.G,
            M: this.M,
            r: this.r,
            H0: this.H0,
            B0: this.B0,
            tau_B: this.tau_B,
            f_TRZ: this.f_TRZ,
            ug1_base: this.ug1_base
        };
    }
}

module.exports = { MagnetarSGR0501_4516 };

File: source14.cpp
/**
 * ================================================================================================
 * UQFF Module 14: Magnetar SGR 0501+4516 MUGE Calculator
 * 
 * SELF-EXPANDING | SELF-UPDATING | SELF-SIMULATING
 * 
 * Description: Master Universal Gravity Equation (MUGE) implementation for SGR 0501+4516 magnetar.
 *              Includes all 12 physics terms: base gravity, cosmic expansion, magnetic decay, Ug,
 *              Lambda, EM, GW, quantum uncertainty, fluid dynamics, oscillatory waves, DM.
 * 
 * Self-Expanding Features:
 *   - Dynamic term registry: Register new PhysicsTerm objects at runtime
 *   - Auto-expanding parameters: Any key accepted, auto-creates on set
 *   - Self-simulation: Run time evolution with automatic data collection
 *   - Self-optimization: Learning rate adjusts terms based on feedback
 *   - Metadata tracking: Creation time, modifications, term history
 * 
 * Unique Physics: Magnetar B(t) field decay and spindown Ω(t) time evolution
 *   B(t) = B₀ × e^(-t/τ_B) - Magnetic field decay
 *   Ω(t) = Ω₀ × e^(-t/τ_Ω) - Rotational spindown
 * 
 * Author: Daniel T. Murphy (daniel.murphy00@gmail.com)
 * Enhanced: January 2026 - Full self-expanding capabilities
 * ================================================================================================
 */

#include "uqff_constants.h"
#include "uqff_self_expanding.h"
#include "uqff_dual_physics.h"
#include <iostream>
#include <iomanip>
#include <cmath>
#include <map>
#include <string>
#include <fstream>
#include <sstream>
#include <vector>
#include <memory>
#include <functional>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

using namespace UQFFExpanding;

// ================================================================================================
// UQFFConfig14 Singleton - Module Configuration
// ================================================================================================
class UQFFConfig14 {
private:
    UQFFConfig14() {
        M = 1.4 * UQFF::SUN_MASS_KG;
        r = 20e3;
        H0 = 2.184e-18;
        B0 = 1e10;
        tau_B = 4000 * 3.156e7;
        B_crit = 1e11;
        Lambda = 1.1e-52;
        q_charge = 1.602e-19;
        v_surf = 1e6;
        f_TRZ = 0.1;
        rho_vac_UA = 7.09e-36;
        rho_vac_SCm = 7.09e-37;
        P_init = 5.0;
        tau_Omega = 10000 * 3.156e7;
        scale_EM = 1e-12;
        proton_mass = 1.673e-27;
        rho_fluid = 1e17;
    }
public:
    double M, r, H0, B0, tau_B, B_crit, Lambda;
    double q_charge, v_surf, f_TRZ, rho_vac_UA, rho_vac_SCm;
    double P_init, tau_Omega, scale_EM, proton_mass, rho_fluid;
    
    static UQFFConfig14& getInstance() {
        static UQFFConfig14 instance;
        return instance;
    }
    UQFFConfig14(const UQFFConfig14&) = delete;
    void operator=(const UQFFConfig14&) = delete;
};

// ================================================================================================
// Custom PhysicsTerm: Magnetar Spindown Term
// ================================================================================================
class SpindownTerm : public PhysicsTerm {
private:
    double tau_Omega;
    double amplitude;
public:
    SpindownTerm(double tau = 3.156e14, double amp = 1e10)
        : tau_Omega(tau), amplitude(amp) {}
    
    double compute(double t, const std::map<std::string, double>& params) const override {
        double omega_decay = std::exp(-t / tau_Omega);
        return amplitude * (1.0 - omega_decay) * omega_decay;
    }
    
    std::string getName() const override { return "Spindown"; }
    std::string getDescription() const override { 
        return "Ω(t) = Ω₀×e^(-t/τ_Ω) rotational spindown energy"; 
    }
    
    void optimize(double learningRate, double error) override {
        amplitude *= (1.0 - learningRate * error * 0.1);
    }
};

// ================================================================================================
// Custom PhysicsTerm: Magnetic Field Decay
// ================================================================================================
class MagneticDecayTerm : public PhysicsTerm {
private:
    double B0, tau_B, B_crit;
public:
    MagneticDecayTerm(double b0 = 1e10, double tau = 1.26e11, double bcrit = 1e11)
        : B0(b0), tau_B(tau), B_crit(bcrit) {}
    
    double compute(double t, const std::map<std::string, double>& params) const override {
        double Bt = B0 * std::exp(-t / tau_B);
        double ug1 = params.count("ug1_base") ? params.at("ug1_base") : 4.65e11;
        return ug1 * (1.0 - Bt / B_crit);
    }
    
    std::string getName() const override { return "MagneticDecay"; }
    std::string getDescription() const override { 
        return "B(t) = B₀×e^(-t/τ_B) magnetic field decay effect"; 
    }
    
    void optimize(double learningRate, double error) override {
        B0 *= (1.0 - learningRate * error * 0.05);
    }
};

// ================================================================================================
// MagnetarSGR0501_4516 - Core Physics Class (12 MUGE Terms)
// ================================================================================================
class MagnetarSGR0501_4516 {
private:
    double M, r, H0, B0, tau_B, B_crit, Lambda;
    double q_charge, v_surf, f_TRZ, rho_vac_UA, rho_vac_SCm;
    double P_init, tau_Omega, scale_EM, proton_mass;
    double t_Hubble, delta_x, delta_p, integral_psi;
    double rho_fluid, A_osc, k_osc, omega_osc, x_pos;
    double M_DM_factor, delta_rho_over_rho;
    double ug1_base;

public:
    MagnetarSGR0501_4516(const UQFFConfig14& cfg) {
        M = cfg.M;
        r = cfg.r;
        H0 = cfg.H0;
        B0 = cfg.B0;
        tau_B = cfg.tau_B;
        B_crit = cfg.B_crit;
        Lambda = cfg.Lambda;
        q_charge = cfg.q_charge;
        v_surf = cfg.v_surf;
        f_TRZ = cfg.f_TRZ;
        rho_vac_UA = cfg.rho_vac_UA;
        rho_vac_SCm = cfg.rho_vac_SCm;
        P_init = cfg.P_init;
        tau_Omega = cfg.tau_Omega;
        scale_EM = cfg.scale_EM;
        proton_mass = cfg.proton_mass;
        rho_fluid = cfg.rho_fluid;
        
        t_Hubble = 13.8e9 * 3.156e7;
        delta_x = 1e-10;
        delta_p = UQFF::hbar / delta_x;
        integral_psi = 1.0;
        A_osc = 1e10;
        k_osc = 1.0 / r;
        omega_osc = 2 * M_PI / P_init;
        x_pos = r;
        M_DM_factor = 0.1;
        delta_rho_over_rho = 1e-5;
        
        ug1_base = (UQFF::G * M) / (r * r);
    }
    
    double B_t(double t) const { return B0 * std::exp(-t / tau_B); }
    double Omega_t(double t) const { return (2 * M_PI / P_init) * std::exp(-t / tau_Omega); }
    double dOmega_dt(double t) const {
        double omega0 = 2 * M_PI / P_init;
        return omega0 * (-1.0 / tau_Omega) * std::exp(-t / tau_Omega);
    }
    
    double compute_Ug(double Bt) const {
        double Ug1 = ug1_base;
        double Ug4 = Ug1 * (1 - Bt / B_crit);
        return (Ug1 + Ug4) * (1 + f_TRZ);
    }
    
    double compute_V() const { return (4.0/3.0) * M_PI * r * r * r; }
    
    // ==================== MAIN MUGE COMPUTATION (12 TERMS) ====================
    double compute_g_MUGE(double t) const {
        if (t < 0) return 0.0;
        
        double Bt = B_t(t);
        double dOdt = dOmega_dt(t);
        double mu_0 = 4 * M_PI * 1e-7;
        double V = compute_V();
        
        // Terms 1-12
        double term_base = ug1_base * (1 + H0 * t) * (1 - Bt / B_crit);
        double term_Ug = compute_Ug(Bt);
        double term_Lambda = (Lambda * UQFF::c * UQFF::c) / 3.0;
        double term_EM = (q_charge * v_surf * Bt / proton_mass) * (1 + rho_vac_UA/rho_vac_SCm) * scale_EM;
        double term_GW = ((UQFF::G * M * M) / (std::pow(UQFF::c, 4) * r)) * (dOdt * dOdt);
        double term_Quantum = (UQFF::hbar / std::sqrt(delta_x * delta_p)) * integral_psi * (2 * M_PI / t_Hubble);
        double term_Fluid = (rho_fluid * V * ug1_base) / M;
        double term_Osc1 = 2 * A_osc * std::cos(k_osc * x_pos) * std::cos(omega_osc * t);
        double term_Osc2 = (2 * M_PI / t_Hubble) * A_osc * std::cos(k_osc * x_pos - omega_osc * t);
        double M_dm = M * M_DM_factor;
        double term_DM = ((M + M_dm) * (delta_rho_over_rho + 3 * UQFF::G * M / (r*r*r))) / M;
        double term_Magnetic = (Bt * Bt) / (mu_0 * rho_fluid * r);
        double dB_dt = -B0 / tau_B * std::exp(-t / tau_B);
        double term_Decay = (Bt * dB_dt) / (mu_0 * rho_fluid * r);
        
        return term_base + term_Ug + term_Lambda + term_EM + term_GW + term_Quantum +
               term_Fluid + term_Osc1 + term_Osc2 + term_DM + term_Magnetic + term_Decay;
    }
    
    double compute_g_Newton() const { return UQFF::G * M / (r * r); }
    double getMass() const { return M; }
    double getRadius() const { return r; }
    double getB0() const { return B0; }
    double getPeriod() const { return P_init; }
    double getUg1Base() const { return ug1_base; }
};

// ================================================================================================
// UQFFModule14 - Self-Expanding Module with full capabilities
// ================================================================================================
class UQFFModule14 : public SelfExpandingModule<UQFFConfig14> {
private:
    MagnetarSGR0501_4516 magnetar;
    
public:
    UQFFModule14()
        : SelfExpandingModule<UQFFConfig14>("UQFFModule14_SGR0501_SelfExpanding"),
          magnetar(UQFFConfig14::getInstance())
    {
        auto& cfg = UQFFConfig14::getInstance();
        params["M"] = cfg.M;
        params["r"] = cfg.r;
        params["B0"] = cfg.B0;
        params["ug1_base"] = magnetar.getUg1Base();
        params["rho_fluid"] = cfg.rho_fluid;
        params["tau_B"] = cfg.tau_B;
        params["tau_Omega"] = cfg.tau_Omega;
        params["G"] = UQFF::G;
        params["hbar"] = UQFF::hbar;
        
        // Register magnetar-specific dynamic terms
        registerDynamicTerm(std::make_unique<SpindownTerm>(cfg.tau_Omega, 1e10));
        registerDynamicTerm(std::make_unique<MagneticDecayTerm>(cfg.B0, cfg.tau_B, cfg.B_crit));
        
        setMetadata("object", "SGR 0501+4516 Magnetar");
        setMetadata("physics_terms", "12 MUGE + dynamic");
    }
    
    double compute(double t) {
        double base_result = magnetar.compute_g_MUGE(t);
        double dynamic_result = computeDynamicTerms(t);
        double total = base_result + dynamic_result;
        
        if (enableLogging) {
            std::cout << "[" << moduleName << "] compute(t=" << std::scientific 
                      << std::setprecision(4) << t << "): base=" << base_result 
                      << ", dynamic=" << dynamic_result << ", total=" << total << "\n";
        }
        return total;
    }
    
    void runSelfSimulation(double t_start, double t_end, int steps) {
        runSimulation(t_start, t_end, steps, [this](double t) { return this->compute(t); });
    }
    
    void printInfo() {
        std::cout << "==========================================================\n";
        std::cout << "  UQFF Module 14: SGR 0501+4516 Magnetar\n";
        std::cout << "  SELF-EXPANDING | SELF-UPDATING | SELF-SIMULATING\n";
        std::cout << "==========================================================\n";
        std::cout << std::scientific << std::setprecision(3);
        std::cout << "M: " << magnetar.getMass() << " kg (1.4 M_sun)\n";
        std::cout << "r: " << magnetar.getRadius() << " m (20 km)\n";
        std::cout << "B0: " << magnetar.getB0() << " T (10 GT)\n";
        std::cout << "\n";
        printExpandedInfo();
    }
    
    const UQFFConfig14& getConfig() const { return UQFFConfig14::getInstance(); }
    double getNewtonianG() const { return magnetar.compute_g_Newton(); }
};

// ================================================================================================
// Main Function - Self-Expanding Demo
// ================================================================================================
int main() {
    std::cout << "==========================================================\n";
    std::cout << "  UQFF Module 14: SGR 0501+4516 Magnetar\n";
    std::cout << "  SELF-EXPANDING | SELF-UPDATING | SELF-SIMULATING\n";
    std::cout << "==========================================================\n\n";
    
    UQFFModule14 module;
    module.setEnableLogging(true);
    module.printInfo();
    
    // SELF-EXPANDING
    std::cout << "\n=== SELF-EXPANDING: Register New Physics Terms ===\n";
    module.registerDynamicTerm(std::make_unique<DynamicVacuumTerm>(1e-10, 1e-14));
    module.registerDynamicTerm(std::make_unique<QuantumCouplingTerm>(1e-38));
    
    std::cout << "\nRegistered Dynamic Terms:\n";
    for (const auto& name : module.listDynamicTerms()) {
        std::cout << "  - " << name << "\n";
    }
    
    // AUTO-EXPANDING PARAMETERS
    std::cout << "\n=== AUTO-EXPANDING: Create New Parameters ===\n";
    module.setDynamicParameter("burst_amplitude", 1e20);
    module.setDynamicParameter("flare_frequency", 0.1);
    
    // SELF-SIMULATION
    std::cout << "\n=== SELF-SIMULATION: Magnetar Evolution ===\n";
    double year = 3.156e7;
    module.runSelfSimulation(0.0, 10000.0 * year, 10);
    
    // STATE PERSISTENCE
    module.exportState("module14_self_expanding_state.txt");
    module.exportSimulation("module14_simulation.csv");
    
    // COMPARISON
    std::cout << "\n=== Comparison ===\n";
    double g_newton = module.getNewtonianG();
    double g_expanded = module.compute(0.0);
    std::cout << "g_Newton: " << std::scientific << std::setprecision(4) << g_newton << " m/s²\n";
    std::cout << "g_Expanded(t=0): " << g_expanded << " m/s²\n";
    std::cout << "Enhancement: " << g_expanded / g_newton << "x\n";
    
    // ==================== DUAL PHYSICS VALIDATION ====================
    std::cout << "\n=== Dual Physics Validation ===\n";
    
    using namespace UQFFDualPhysics;
    
    // Initialize FluidSolver
    FluidSolver fluidSolver(32, 0.1, 0.0001);
    fluidSolver.add_jet_force(10.0);
    for (int step = 0; step < 10; ++step) {
        fluidSolver.step(g_expanded * 1e-10);
    }
    std::cout << "FluidSolver: Max velocity = " << fluidSolver.getMaxVelocity() << " m/s\n";
    
    // Initialize DualMethodValidator
    auto& cfg = module.getConfig();
    DualMethodValidator validator("source14_dual_physics.log");
    validator.addConstraint("SGR0501+4516", 1e10, 1e14, 15.0);
    
    UQFFDualPhysics::CelestialBody body14("SGR0501+4516", cfg.M, cfg.r, cfg.B0);
    UQFFDualPhysics::MUGESystem muge14("SGR0501+4516", cfg.M, cfg.r);
    muge14.B0 = cfg.B0;
    muge14.Lambda = cfg.Lambda;
    
    auto result = validator.validate(body14, muge14, 0.0, 0.0);
    result.print();
    std::cout << "Dual Physics: IMPLEMENTED\n";
    // ================================================================
    
    std::cout << "\n[Module14] Self-expanding demonstration complete.\n";
    return 0;
}
